### ISSUE TITLE
koreader should support other OPDS rel="..." links

### ISSUE THREAD

**Issue by mikespub** [koreader/koreader#10871]
See https://github.com/mikespub-org/seblucas-cops/issues/26#issuecomment-1703806763 as discovered by @Mossop

If the OPDS feed contains a link with a rel="..." it doesn't explicitly recognize, koreader fails to follow the link and causes "Cannot get catalog info from nil".

```
<link href="feed.php?page=6" type="application/atom+xml;profile=opds-catalog;kind=navigation"/> OK
<link href="feed.php?page=6" type="application/atom+xml;profile=opds-catalog;kind=navigation" rel="subsection"/> OK
<link href="feed.php?page=6" type="application/atom+xml;profile=opds-catalog;kind=navigation" rel="section"/> Not OK
```

Apparently this might be due to the following code section: https://github.com/koreader/koreader/blob/a767ad44db5a22c6d14c505241e8698fa3113c69/plugins/opds.koplugin/opdsbrowser.lua#L379

koreader should be able to handle the case where link.rel contains something it doesn't know about too, since the OPDS specs "encourage" the use of any relation defined in RFC5988...

**Comment by NiLuJe** [comment]
Could you quote the specs on that?

I don't see anything on a (very) quick skim of https://specs.opds.io/opds-1.2

**Comment by NiLuJe** [comment]
A full testcase might be helpful, too, FWIW.

**Comment by NiLuJe** [comment]
I'm sort of fine with ignoring unknown stuff by default, though?

(Admittedly, the error reporting kinda sucks in that case, but that's nothing really new ;p).

**Comment by mikespub** [comment]
I don't know anything about the koreader code base - or lua for that matter :-)

But if you have a test case parsing some OPDS feed somewhere, it should be fairly easy to add a link with an unknown rel="..." like the example above, and check what happens no?

**Comment by Frenzie** [comment]
The question is if that's actually a realistic (or prudent) scenario, not what would happen in that case. If a thumbnail rel were unknown and treated as just a synonym for subsection due to being unknown, things may well go wrong for example.

Put another way, `section` and any random old string are not the same thing.

**Comment by mikespub** [comment]
Agreed, but in principle any of the relations defined in https://datatracker.ietf.org/doc/html/rfc5988#section-6.2.2 could be present in an OPDS feed. You already have self, start, prev, next, first, last, etc. but also potentially subsection, related, and so on.

So as long as it doesn't match any "known" relation that you want to treat in a special way in your application, you should treat it as a generic link - not break down in agony and say "Cannot get catalog info from nil" like it does now apparently :-)

**Comment by NiLuJe** [comment]
It's been a while since I dealt with this, but it's my understanding that the complete and utter breakdown should only happen it it finds *nothing* to handle (at all), which is not the same thing as finding a few unexpected or unhandled rels *in addition* to known stuff.

**Comment by mikespub** [comment]
Ah - my misunderstanding, sorry. From what @Mossop explained in his original issue, the initial feed page loaded with links to the different sections, but when clicking on one of the sections, it gave that error message.

For example like https://owncloud.mikespub.net/cops/feed.php - but with rel="section" instead of rel="subsection" in each link

**Comment by Frenzie** [comment]
> So as long as it doesn't match any "known" relation that you want to treat in a special way in your application, you should treat it as a generic link - not break down in agony and say "Cannot get catalog info from nil" like it does now apparently :-)

The spec's wording seems to have changed in 1.2 ftr, see 1.1 which the implementation was based on here:
https://specs.opds.io/opds-1.1.html#OPDS_Catalog_Relations

Here's the reasoning behind what it does:
https://github.com/koreader/koreader/pull/1300#issuecomment-64951892

> The original rationale was to ignore "facet" links that koreader does not know how to handle.
According to the [the spec](http://opds-spec.org/specs/opds-catalog-1-1-20110627/#Facets), entries may multiple links of type application/atom+xml;profile=opds-catalog. One of them is the main link we are interested in, the others are so-called "facets", that provide some additional view into the same data. The current code does not support entries with multiple links, so our best choce for now is to ignore those "facets" and use the main link (with no rel attribute).

Note my reply:
> I think I might prefer blacklisting that which is known not to work over blacklisting that which might work, but I suppose either way is fine.

Also see prior art doing the same thing: https://github.com/geometer/FBReaderJ/blob/d73f7816cf5a1cf4280a4a676143b4dda0f39e3e/src/org/geometerplus/fbreader/network/opds/OPDSLinkXMLReader.java#L71

**Comment by NiLuJe** [comment]
> I think I might prefer blacklisting that which is known not to work over blacklisting that which might work, but I suppose either way is fine.

Right, so listing that `section` entry was by design at least, and we're mostly back to "the error reporting sucks" ^^.

**Comment by Frenzie** [comment]
Yes, fundamentally the only real problem as the code was intended is that it landed on that `nil` thing, though given the changed OPDS 1.2 spec it might be the case that circumstances have changed slightly too.

**Comment by mikespub** [comment]
For what it's worth, I tested with the old FBReader as well, and the main difference I can see is that it simply ignores the rel="..." part it doesn't know, but it still shows a working link for catalogs as it should.

koreader may have had issues with "facet" links in acquisition feeds, but other relation links were just as valid in OPDS 1.1 and OPDS 1.2 feeds.

So if you decide to show them and ignore the rel="..." that's fine, or if you decide not to show them that's fine too - but showing them & having it cause a `nil` error, that's probably not a good combination ;-)

Either way, it's your app, so your choice - in the meantime I reverted the rel="..." that caused problems in COPS with [release 1.4.4](https://github.com/mikespub-org/seblucas-cops/releases/tag/1.4.4)

**Comment by PiperKev** [comment]
Has this ever been fixed? I get the  "Cannot get catalog info from nil" error when I try to access the Standard Ebooks "All Standard Ebooks" OPDS feed.
